### PR TITLE
Implementing Escape State Machine for process console and command navigations

### PR DESCRIPTION
**Pull request by mihai-negru**
### Thanks

Thanks to [jettr](https://github.com/jettr) for the escape state machine code snapshot and feedback

### Pull Request Overview
This pull request implements the jettr's suggestion of implementing an escape state machine (https://github.com/tock/tock/pull/3381#discussion_r1096079934).

Also this pull request implements new features over typing experience in the `ProcessConsole`, in order to make typing more dynamic. 

### Added features
* Use `Left` and `Right` arrows in order to navigate through a typing command.
* Use `Home` and `End` commands to move at the beginning or end of a command.
* Press `Delete` key to remove the character under the cursor.
* Inserting or deleting(using backspace) a character inside a command is now possible.
* Pressing enter inside a command is also possible.

### Flash size for IMIX binary build:
| Builds | text | data | bss | dec |
| ------ | ---- | ----- | ---- | ---- |
| Tock Master | 175592  | 0 | 30108 | 205700 |
| New Code Size | 176232  | 0 | 30116 | 206348 |
| Differences | 640 | 0 | 8 | 648 | 

* CommandHistory is enabled by default on Tock Master, so the size compiled refers to `new_code + history_size`

### Testing Strategy

This pull request was tested on a `nrf52840dk` board

### Changed Files
`capsules/core/src/process_console.rs`
* Implementation for new features.  

### Documentation Updated

Updated the documentation in `doc/Process_Console.md` accordingly:
* Support on how to use the new features.
* Explaining that whitespaces do not affect the resulting command.

### TODO or Help Wanted

Feedback relating to optimize the new features and how to make a better `command_history`.

### Formatting

- [x] Ran `make prepush`.

### CODE REVIEW

**Comment by mihai-negru** [comment]
> ### Pull Request Overview
> This pull request implements the jettr's suggestion of implementing an escape state machine ([tock#3381 (comment)](https://github.com/tock/tock/pull/3381#discussion_r1096079934))
> 
> The state machine proposed by jettr, has many functionalities, however for this pull request were implemented scrolling in `command_history` maintaining the same behavior as the last merged PR ([tock#3381](https://github.com/tock/tock/pull/3381))
> 
> #### Scketch pull request
> The pull request is not finished according to jettr's suggestion, because the functionalities of home, delete, end, left and right are not implemented.
> 
> ### Testing Strategy
> This pull request was tested on `nrf52840dk` board
> 
> ### TODO or Help Wanted
> Suggestions in order to make a better `command_history`
> 
> ### Documentation Updated
> * N/A
> 
> ### Formatting
> * [x]  Ran `make prepush`.

About the suggestions, I would like to modify the `command_history`:
- Translate the array in a structure containing an array of Commands
- Encapsulate any functionality in the history structure
- Some functionalities require code duplication (at up and down match branch), which can be solved by creating a history structure, that will keep track of modified history and last modified byte (the byte before escape machine started)
- After encapsulating the array into a structure, we may focus on implementing the left and right functionalities
- To many command history functionalities are maintained by process console and not by a different structure

I would like to know what do you think about this perspective?

**Comment by mihai-negru** [comment]
I have changed the code a bit in order to encapsulate the history functionality in another structure, however I have not put the **existance** of the escape machine under the **COMMAND_HISTORY_LEN > 1** condition, because for futher implementations for Left and Right functionalities we will not be needed the COMMANS_HISTORY_LEN. Also if you disagree with the encapsulation we can forget about this and go back to the commit `#765cc43`.

When we will implement the Left, Right, Home and End, keys the code will be changed a little bit more because we have to get rid of the `if let Some(index) = .....`, so by encapsulating, we somehow detached the ProcessConsole from the CommandHistory, so it will be easier to make the futher implementations.

**Comment by mihai-negru** [comment]
Also in jettr's message he metioned something about a `mru`, could you redirect me to a website or some papers in order to understand how to use this and how to implement it in our PR, thanks in advice

**Comment by mihai-negru** [comment]
For this implementation the size has increased around 200 bytes in both enabled and disabled states, because the escape machine is not designed in general just for the command history

**Comment by mihai-negru** [comment]
Kind reminder to review this, I need this in order to continue with further implementations. :)

File: capsules/src/process_console.rs [hunk 1]
```diff
@@ -75,6 +75,65 @@ impl Default for WriterState {
...
+    Up,
+    Down,
+    Left,
+    Right,
+}
+
+/// Escape state machine to check if
+/// an escape sequence has occured
+#[derive(Copy, Clone)]
+enum EscState {
```
Comment (alexandruradovici) [inline capsules/src/process_console.rs]: Please add a description for all the items.

File: capsules/src/process_console.rs [hunk 2]
```diff
@@ -202,6 +257,85 @@ impl PartialEq<[u8; COMMAND_BUF_LEN]> for Command {
...
+    /// Checks if the command line was modified
+    /// before pressing Up or Down keys
+    /// and saves the current modified command
+    /// into the history
+    fn change_cmd_from(&mut self, cmd: &[u8]) {
+        match self.modified_byte {
+            b' ' => {
+                self.cmds[0].delete_last_byte();
+            }
+            b'\x08' | b'\x7F' => {
```
Comment (alexandruradovici) [inline capsules/src/process_console.rs]: Please define constants

File: capsules/src/process_console.rs [hunk 3]
```diff
@@ -950,23 +1076,70 @@ impl<'a, const COMMAND_HISTORY_LEN: usize, A: Alarm<'a>, C: ProcessManagementCap
...
+
+                                            let prev_command_len = self.command_index.get();
+                                            let next_command_len = ht.cmds[index].len;
+
+                                            // Clear the displayed command
+                                            for _ in 0..prev_command_len {
+                                                let _ = self.write_bytes(&[
+                                                    '\x08' as u8,
+                                                    ' ' as u8,
+                                                    '\x08' as u8,
```
Comment (alexandruradovici) [inline capsules/src/process_console.rs]: Please define a constant for this.

File: capsules/src/process_console.rs [hunk 4]
```diff
@@ -984,127 +1157,18 @@ impl<'a, const COMMAND_HISTORY_LEN: usize, A: Alarm<'a>, C: ProcessManagementCap
...
                                     });
                                 }
                             }
-                        } else if (COMMAND_HISTORY_LEN > 1)
-                            && (read_buf[0] == ESC || self.control_seq_in_progress.get())
+                        } else if (COMMAND_HISTORY_LEN > 1) && (esc_state.has_started()) {
+                            self.command_history
+                                .map(|ht| ht.modified_byte = previous_byte);
+                        } else if index < (command.len() - 1)
+                            && read_buf[0] < 128
```
Comment (alexandruradovici) [inline capsules/src/process_console.rs]: 128?

File: capsules/src/process_console.rs [hunk 5]
```diff
@@ -950,23 +1076,70 @@ impl<'a, const COMMAND_HISTORY_LEN: usize, A: Alarm<'a>, C: ProcessManagementCap
...
+                                        } {
+                                            ht.change_cmd_from(&command);
+
+                                            let prev_command_len = self.command_index.get();
+                                            let next_command_len = ht.cmds[index].len;
+
+                                            // Clear the displayed command
+                                            for _ in 0..prev_command_len {
+                                                let _ = self.write_bytes(&[
+                                                    '\x08' as u8,
```
Comment (alexandruradovici) [inline capsules/src/process_console.rs]: Please define a constant for this.

**Review by alexandruradovici — CHANGES_REQUESTED** [review]
Looks really nice.

File: capsules/src/process_console.rs (hunk 4 above)
Comment (mihai-negru) [reply capsules/src/process_console.rs]: This was not added by me, I can't say what is it used for.

File: capsules/src/process_console.rs (hunk 2 above)
Comment (mihai-negru) [reply capsules/src/process_console.rs]: Done.

File: capsules/src/process_console.rs (hunk 3 above)
Comment (mihai-negru) [reply capsules/src/process_console.rs]: Done.

File: capsules/src/process_console.rs (hunk 5 above)
Comment (mihai-negru) [reply capsules/src/process_console.rs]: Done.

File: capsules/src/process_console.rs (hunk 1 above)
Comment (mihai-negru) [reply capsules/src/process_console.rs]: Done.

File: capsules/src/process_console.rs (hunk 4 above)
Comment (mihai-negru) [reply capsules/src/process_console.rs]: I have inspected the `File Blame` using the git lens and seen that `read_buf[0] < 128` was added by the creator of the ProcessConsole `Philips Levis` with the following explanation:

```text
// For some reason, sometimes reads return > 127 but no error,
// which causes utf-8 decoding failure, so check byte is < 128. -pal
```

However, I created a constant for 128, in order not to make further confusions.

**Comment by mihai-negru** [comment]
With this new commit, I have solved the review problems.

As usually, I have run `make prepush` and tested the modification on the boards in order to see if the program keeps working as expected.

With your confirmation I would like to move on the next phase, in order to implement `Left and Right` functionalities.

So the next phases that I have in mind are:
* Left and Right implementation (phase 1)
* Home and End implementation (phase 2)
* Delete and Insert implementation (phase 3)
* Clear screen command (phase 4)

Please, I would like to know if it is alright for the next steps.

Thanks for the review.

**Comment by alexandruradovici** [comment]
Yes, the PR looks very nice, start implementing the rest.

**Comment by mihai-negru** [comment]
With these commits, I have implemented the left, right, home, end and delete functionalities.

As usual I have tested the implementation on a nrf52840dk board.

I have also signed some constants to `\r` and `\n` characters.

**Comment by mihai-negru** [comment]
Moving to implement the `clear` function for process console in order to clear the screen

**Comment by mihai-negru** [comment]
As I have seen so far the ProcessConsole does not support escape sequences, because there are several ways to clear the screen using some escape sequences for example `printf '\033c'`, is there a way in order to send the following bytes `\033c` to the host terminal in order to handle them.

Another way of implementing the `clear` functionality, would be to print a bunch of `\r\n`, however it lefts a bunch of space between the top of the terminal and the command line.

Is there a way to achieve this, by using escape sequences?

Thanks for the answer.

**Comment by mihai-negru** [comment]
This PR was send successfully to the upstream: https://github.com/tock/tock/pull/3414

**Comment by alexandruradovici** [comment]
@mihai-negru any updates here?

**Comment by mihai-negru** [comment]
> @mihai-negru any updates here?

Hi, sorry, I was a bit busy with math olympics, I am getting back to work this week in order to create the test script file for the escape state macilhine

**Comment by mihai-negru** [comment]
Hi, this pull request was merged into tock master at https://github.com/tock/tock/pull/3414

**Pull request merged**